helix-core/src/main/java/org/apache/helix/monitoring/mbeans/InstanceMonitor.java

```diff
@@ -24,35 +24,86 @@
 import java.util.Map;
 import java.util.Set;
 
+import javax.management.JMException;
+import javax.management.ObjectName;
+
 import com.google.common.base.Joiner;
 import com.google.common.collect.ImmutableList;
 import com.google.common.collect.Lists;
+import org.apache.helix.monitoring.mbeans.dynamicMBeans.DynamicMBeanProvider;
+import org.apache.helix.monitoring.mbeans.dynamicMBeans.DynamicMetric;
+import org.apache.helix.monitoring.mbeans.dynamicMBeans.SimpleDynamicMetric;
+
 
 /**
  * Implementation of the instance status bean
  */
-public class InstanceMonitor implements InstanceMonitorMBean {
+public class InstanceMonitor extends DynamicMBeanProvider {
+  /**
+   * Metric names for instance capacity.
+   */
+  public enum InstanceMonitorMetrics {
+    TOTAL_MESSAGE_RECEIVED_COUNTER("TotalMessageReceivedCounter"),
+    ENABLED_STATUS_GAUGE("EnabledStatusGauge"),
+    ONLINE_STATUS_GAUGE("OnlineStatusGauge"),
+    MAX_CAPACITY_USAGE_GAUGE("MaxCapacityUsageGauge"),
+    DISABLED_PARTITIONS_GAUGE("DisabledPartitionsGauge");
+
+    private String metricName;
+
+    InstanceMonitorMetrics(String name) {
+      metricName = name;
+    }
+
+    public String metricName() {
+      return metricName;
+    }
+  }
+
   private final String _clusterName;
   private final String _participantName;
   private List<String> _tags;
-  private long _disabledPartitions;
-  private boolean _isUp;
-  private boolean _isEnabled;
-  private long _totalMessageReceived;
+  private ObjectName _initObjectName;
+
+  // Counters
+  private SimpleDynamicMetric<Long> _totalMessagedReceivedCounter;
+
+  // Gauges
+  private SimpleDynamicMetric<Long> _enabledStatusGauge;
+  private SimpleDynamicMetric<Long> _disabledPartitionsGauge;
+  private SimpleDynamicMetric<Long> _onlineStatusGauge;
+  private SimpleDynamicMetric<Double> _maxCapacityUsageGauge;
 
   /**
    * Initialize the bean
    * @param clusterName the cluster to monitor
    * @param participantName the instance whose statistics this holds
    */
-  public InstanceMonitor(String clusterName, String participantName) {
+  public InstanceMonitor(String clusterName, String participantName, ObjectName objectName)
+      throws JMException {
     _clusterName = clusterName;
     _participantName = participantName;
     _tags = ImmutableList.of(ClusterStatusMonitor.DEFAULT_TAG);
-    _disabledPartitions = 0L;
-    _isUp = false;
-    _isEnabled = false;
-    _totalMessageReceived = 0;
+    _initObjectName = objectName;
+
+    createMetrics();
+    register();
```

Comment:
To follow our convention, please don't register in the constructor. Trigger the register after the instantiation in the caller method.

Reply:
OK. Maybe because there are some mbeans registered in the constructor.